File: src/ios/XUPPayExt.m

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "XUPPayExt.h"
#import <Cordova/CDVInvokedUrlCommand.h>
#import <Cordova/CDVPluginResult.h>
#import <XFace/XConstants.h>

@implementation XUPPayDelegate

- (id) initWithCallbackId:(NSString *)callbackId
{
    self = [super init];
    if (self)
    {
        self->_callbackId = callbackId;
    }
    return self;
}

- (void)UPPayPluginResult:(NSString *)result
{
    /** result:
        支付成功：银行卡卡号
        支付失败：fail
        支付取消：cancel
     */
    CDVPluginResult *extResult;
    if([result isEqualToString:@"fail"] || [result isEqualToString:@"cancel"]){
        extResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:result];
    } else {
        extResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"success"];
    }
    [self.uppayExt.commandDelegate sendPluginResult:extResult callbackId:self->_callbackId];

    if(IsAtLeastiOSVersion(@"7.0"))
    {
        //在iOS 7.0上，还原status bar显示状态
        [[UIApplication sharedApplication] setStatusBarHidden:self.statusBarHidden];
    }

    [self.uppayExt removeUPPayDelegate:self];
    self.uppayExt = nil;
}

/**
    @成功调用控件以后，会执行回调函数。该函数返回一个字典，下面是关键字的意义：

    @key:result                value：(success,fail,cancel)
    @key:balance               value：账面余额（仅在成功时返回，其余为nil）
    @key:availableBalance      value：可用余额（仅在成功时返回，其余为nil）

 */
- (void)balanceQueryResult:(NSDictionary *)resultDict
{
    CDVPluginResult *extResult;
    NSString* result = [resultDict objectForKey:@"result"];
    if([result isEqualToString:@"success"])
    {
        extResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:resultDict];
    }
    else
    {
        extResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    [self.uppayExt.commandDelegate sendPluginResult:extResult callbackId:self->_callbackId];

    if(IsAtLeastiOSVersion(@"7.0"))
    {
        //在iOS 7.0上，还原status bar显示状态
        [[UIApplication sharedApplication] setStatusBarHidden:self.statusBarHidden];
    }
    [self.uppayExt removeUPPayDelegate:self];
    self.uppayExt = nil;
}

@end

@implementation XUPPayExt

- (CDVPlugin*)initWithWebView:(UIWebView *)theWebView
{
    self = [super initWithWebView:theWebView];
    if (self)
    {
        self->uppayDelegates = [[NSMutableArray alloc] init];
    }

    return self;
}

- (void)startPay:(CDVInvokedUrlCommand *)command
{
    NSString *transSerialNumber = [command.arguments objectAtIndex:0];
    //"00" for formal environment, "01" for test environment
    NSString *mode = [command.arguments objectAtIndex:1];
    NSString *sysProvide = [command.arguments objectAtIndex:2];
    NSString *spId = [command.arguments objectAtIndex:3];
    NSArray *cards = [command argumentAtIndex:4 withDefault:[NSArray array]];
    XUPPayDelegate *delegate = [[XUPPayDelegate alloc] initWithCallbackId:command.callbackId];

    delegate.statusBarHidden = [UIApplication sharedApplication].statusBarHidden;
    if(IsAtLeastiOSVersion(@"7.0"))
    {
        // iOS 7.0上，使用全屏方式启动支付控件，解决状态栏与支付控件内容发生重叠的问题
        [[UIApplication sharedApplication] setStatusBarHidden:YES];
    }

    if ([UPPayPlugin respondsToSelector:@selector(startPay:sysProvide:spId:mode:viewController:delegate:)])
    {
        [UPPayPlugin startPay:transSerialNumber sysProvide:sysProvide spId:spId mode:mode viewController:[self viewController] delegate:delegate];
    }
    else if ([UPPayPlugin respondsToSelector:@selector(startPay:mode:viewController:delegate:)])
    {
        [UPPayPlugin startPay:transSerialNumber mode:mode viewController:[self viewController] delegate:delegate];
    }
    else if ([UPPayPlugin respondsToSelector:@selector(startPay:mode:viewController:delegate:cards:)])
    {
        [UPPayPlugin startPay:transSerialNumber mode:mode viewController:[self viewController] delegate:delegate cards:cards];
    }

    [uppayDelegates addObject:delegate];
    delegate.uppayExt = self;
}

- (void)startBalanceEnquire:(CDVInvokedUrlCommand *)command
{
    NSString *pan = [command.arguments objectAtIndex:0];
    //"00" for formal environment, "01" for test environment
    NSString *mode = [command.arguments objectAtIndex:1];
    XUPPayDelegate *delegate = [[XUPPayDelegate alloc] initWithCallbackId:command.callbackId];

    delegate.statusBarHidden = [UIApplication sharedApplication].statusBarHidden;
    if(IsAtLeastiOSVersion(@"7.0"))
    {
        // iOS 7.0上，使用全屏方式启动支付控件，解决状态栏与支付控件内容发生重叠的问题
        [[UIApplication sharedApplication] setStatusBarHidden:YES];
    }

    BOOL ret = [UPBalanceQueryPlugin startQueryBalanceWithPan:pan  mode:mode container:[self viewController] delegate:delegate];

    if (!ret) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [uppayDelegates addObject:delegate];
    delegate.uppayExt = self;
}

- (void) removeUPPayDelegate:(XUPPayDelegate *)delegate
{
    [uppayDelegates removeObject:delegate];
}

@end

File: src/android/XUPPayExt.java

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.polyvi.xface.extension.uppay;

import java.util.Locale;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.polyvi.xface.util.XLog;
import com.unionpay.mobile.plugin.zl.UPLauncher;
import com.unionpay.mobile.uppay.v2.UPLauncherV2;

import android.content.Intent;


/**
 * XUPPayExt主要是封装启动银联支付控件的接口和接收交易结束后控件返回的交易结果<br/>
 * dependent-libs: UPPayPluginEx.jar,UPPluginWidget.jar,armeabi/libCDZL.so,armeabi/libCdzlWidget.so
 */
public class XUPPayExt extends CordovaPlugin {
    /**
     * 1.该扩展需要在AndroidManifest.xml配置<activity>标签
     * <activity android:name="com.unionpay.mobile.uppay.v2.MobileActivityEx"
     *      android:configChanges="orientation|keyboardHidden"
     *      android:screenOrientation="portrait"
     *      android:theme="@style/Transparent"
     *      android:windowSoftInputMode="adjustResize" />
     * <activity android:name="com.unionpay.mobile.plugin.zl.MobileActivity"
     *      android:configChanges="orientation|keyboardHidden"
     *      android:screenOrientation="portrait"
     *      android:theme="@style/Transparent"
     *      android:windowSoftInputMode="adjustResize" />
     * 2.该扩展需要在AndroidManifest.xml配置的权限有
     * <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
     * <uses-permission android:name="android.permission.INTERNET" />
     * <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
     * <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     * <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
     * <uses-permission android:name="android.permission.READ_PHONE_STATE" />
     * 3.在res/values中增加style.xml文件
     */
    private static final String CLASS_NAME = XUPPayExt.class.getSimpleName();
    private static final String KEY_PAY_RESULT = "pay_result";
    private static final String PAY_RESULT_SUCCESS = "success";

    private static final String COMMAND_START_PAY = "startPay";
    private static final String COMMAND_START_BALANCE_ENQUIRE = "startBalanceEnquire";

    /**
     * 银联支付控件或者无卡余额查询控件启动时所使用的requestCode，与UPPayPluginEx.jar和UPPluginWidget.jar中的实际值一致
     */
    private static final int UPPAY_PLUGIN_REQUEST_CODE = 100;

    private CallbackContext mCallbackCtx;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        cordova.setActivityResultCallback(this);
        super.initialize(cordova, webView);
    }

    @Override
    public boolean execute(String action, JSONArray args,
            CallbackContext callbackContext) throws JSONException {
        // 目前不支持同时进行多个支付操作
        mCallbackCtx = callbackContext;
        if (COMMAND_START_PAY.equals(action)) {
            String transSerialNumber = args.getString(0);
            // "00"：代表接入生产环境（正式版本需要）
            // "01"：代表接入开发测试环境（测试版本需要）
            // "99": 代表pm测试环境（测试版本需要）
            String mode = args.getString(1);
            JSONArray cards = null;
            if (!args.isNull(4)) {
                cards = args.getJSONArray(4);
            }
            // 保留域
            String reserved = null;
            UPLauncher.startUPPay(cordova.getActivity(),
                    transSerialNumber, cards, mode, reserved);
        } else if (COMMAND_START_BALANCE_ENQUIRE.equals(action)) {
            String pan = args.getString(0);
            // "00":银联生产环境
            // "01":银联测试环境
            String mode = args.getString(1);
            // 保留域
            String reserved = null;
            UPLauncherV2.startBalanceEnquire(cordova.getActivity(), pan,
                    mode, reserved);
        } else {
            return false;
        }
        return true;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
        String payResult = "fail";
        if (intent != null) {
            if (UPPAY_PLUGIN_REQUEST_CODE == requestCode) {
                payResult = intent.getExtras().getString(KEY_PAY_RESULT)
                        .toLowerCase(Locale.ENGLISH);
                if (null != payResult) {
                    if (payResult.equals(PAY_RESULT_SUCCESS)) {
                        mCallbackCtx.success(PAY_RESULT_SUCCESS);
                    } else {
                        mCallbackCtx.error(payResult);
                    }
                } else {
                    // 余额
                    String balanceAmount = intent.getExtras().getString(
                            "balance_amount");
                    // 可用余额
                    String balanceAvailableAmount = intent.getExtras().getString(
                            "balance_available_amount");
                    JSONObject jsonObj = null;
                    if (null == balanceAmount && null == balanceAvailableAmount) {
                        XLog.e(CLASS_NAME, "Return params: balanceAmount or balanceAvailableAmount is null");
                        mCallbackCtx.error(payResult);
                    } else {
                        jsonObj = new JSONObject();
                        try {
                            jsonObj.put("balance", balanceAmount);
                            jsonObj.put("availableBalance", balanceAvailableAmount);
                            mCallbackCtx.success(jsonObj);
                        } catch (JSONException e) {
                            XLog.e(CLASS_NAME, "onActivityResult put element into jsonObj error!");
                            e.printStackTrace();
                            mCallbackCtx.error(payResult);
                        }
                    }
                }
            }
        }
        mCallbackCtx.error(payResult);
    }
}
